system/default_ad_api/src/vehicle.hpp

```diff
@@ -0,0 +1,113 @@
+
+private:
+  using GearReport = vehicle_interface::GearStatus::Message;
+  using ApiGear = autoware_adapi_v1_msgs::msg::Gear;
+  using TurnIndicatorsReport = vehicle_interface::TurnIndicatorStatus::Message;
+  using ApiTurnIndicator = autoware_adapi_v1_msgs::msg::TurnIndicators;
+  using HazardLightsReport = vehicle_interface::HazardLightStatus::Message;
+  using ApiHazardLight = autoware_adapi_v1_msgs::msg::HazardLights;
+  using MapProjectorInfo = vehicle_interface::MapProjectorInfo::Message;
+
+  std::unordered_map<uint8_t, uint8_t> gear_type_ = {
+    {GearReport::NONE, ApiGear::UNKNOWN},    {GearReport::NEUTRAL, ApiGear::NEUTRAL},
+    {GearReport::DRIVE, ApiGear::DRIVE},     {GearReport::DRIVE_2, ApiGear::DRIVE},
+    {GearReport::DRIVE_3, ApiGear::DRIVE},   {GearReport::DRIVE_4, ApiGear::DRIVE},
+    {GearReport::DRIVE_5, ApiGear::DRIVE},   {GearReport::DRIVE_6, ApiGear::DRIVE},
+    {GearReport::DRIVE_7, ApiGear::DRIVE},   {GearReport::DRIVE_8, ApiGear::DRIVE},
+    {GearReport::DRIVE_9, ApiGear::DRIVE},   {GearReport::DRIVE_10, ApiGear::DRIVE},
+    {GearReport::DRIVE_11, ApiGear::DRIVE},  {GearReport::DRIVE_12, ApiGear::DRIVE},
+    {GearReport::DRIVE_13, ApiGear::DRIVE},  {GearReport::DRIVE_14, ApiGear::DRIVE},
+    {GearReport::DRIVE_15, ApiGear::DRIVE},  {GearReport::DRIVE_16, ApiGear::DRIVE},
+    {GearReport::DRIVE_17, ApiGear::DRIVE},  {GearReport::DRIVE_18, ApiGear::DRIVE},
+    {GearReport::REVERSE, ApiGear::REVERSE}, {GearReport::REVERSE_2, ApiGear::REVERSE},
+    {GearReport::PARK, ApiGear::PARK},       {GearReport::LOW, ApiGear::LOW},
+    {GearReport::LOW_2, ApiGear::LOW},
+  };
+
+  std::unordered_map<uint8_t, uint8_t> turn_indicator_type_ = {
+    {TurnIndicatorsReport::DISABLE, ApiTurnIndicator::DISABLE},
+    {TurnIndicatorsReport::ENABLE_LEFT, ApiTurnIndicator::LEFT},
+    {TurnIndicatorsReport::ENABLE_RIGHT, ApiTurnIndicator::RIGHT},
+  };
+
+  std::unordered_map<uint8_t, uint8_t> hazard_light_type_ = {
+    {HazardLightsReport::DISABLE, ApiHazardLight::DISABLE},
+    {HazardLightsReport::ENABLE, ApiHazardLight::ENABLE},
+  };
+
+  rclcpp::CallbackGroup::SharedPtr group_cli_;
+  Pub<autoware_ad_api::vehicle::VehicleKinematics> pub_kinematics_;
+  Pub<autoware_ad_api::vehicle::VehicleStatus> pub_status_;
+  Sub<vehicle_interface::KinematicState> sub_kinematic_state_;
+  Sub<vehicle_interface::Acceleration> sub_acceleration_;
+  Sub<vehicle_interface::SteeringStatus> sub_steering_;
+  Sub<vehicle_interface::GearStatus> sub_gear_state_;
+  Sub<vehicle_interface::TurnIndicatorStatus> sub_turn_indicator_;
+  Sub<vehicle_interface::HazardLightStatus> sub_hazard_light_;
+  Sub<vehicle_interface::MapProjectorInfo> sub_map_projector_info_;
+  Sub<vehicle_interface::EnergyStatus> sub_energy_level_;
+  rclcpp::TimerBase::SharedPtr timer_;
+
+  MapProjectorInfo::ConstSharedPtr map_projector_info_;
+  autoware_ad_api::vehicle::VehicleKinematics::Message vehicle_kinematics_;
```

Comment:
For kinematics, it seems better to save kinematic_state and acceleration and convert them in a timer callback.

Reply:
so I will just save all the msg as pointer and then only process in the callback?